### PR TITLE
[Compiler-v2] Fix issue 12309

### LINKED ISSUE #12309: [Bug][compiler-v2] Missing drop ability check when value is unused
# 🐛 Bug

Complier v2 fails to report a missing drop ability in the following code (whereas compiler v1 does report it), causing bytecode verification error.

```
module 0xc0ffee::m {
    public fun consume<T>(_x: T) {}
}
```

### PR DESCRIPTION
**Pull request by rahxephon89**
### Description

This PR closes #12309 by adding a check to the ability processor that if 1) an unused parameter is of generic type or a struct or a vector of struct; 2) the struct/generic type has no drop ability and 3) the function does not always abort, an error will be generated.

### Test Plan

Added a new test `unused_para_no_drop` to validate the fix

### CODE REVIEW

File: third_party/move/move-compiler-v2/tests/ability-check/unused_para_no_drop.move
Comment (brmataptos) [inline third_party/move/move-compiler-v2/tests/ability-check/unused_para_no_drop.move:1]: Please add test cases where value is used but not consumed, e.g.:
```
public fun f(x: S, y: S): bool {
    &x == &y
}

public fun f<T>(x: T, y: T): bool {
    &x == &y
}

struct S2 {
   foo: u64
}

public fun f(x: S2, y: S2): bool {
    x.foo == y.foo
}
```
and make sure those are also handled.

File: third_party/move/move-compiler-v2/tests/ability-check/explicit_move.move
Comment (rahxephon89) [inline third_party/move/move-compiler-v2/tests/ability-check/explicit_move.move:3]: Add this so that the error message does not change.

File: third_party/move/move-compiler-v2/tests/ability-check/inferred_copy.move
Comment (rahxephon89) [inline third_party/move/move-compiler-v2/tests/ability-check/inferred_copy.move:3]: Add this so that the error message does not change.

File: third_party/move/move-compiler-v2/tests/ability-check/unused_para_no_drop.move
Comment (rahxephon89) [reply third_party/move/move-compiler-v2/tests/ability-check/unused_para_no_drop.move:1]: done

File: third_party/move/move-compiler-v2/src/pipeline/ability_processor.rs
Comment (wrwg) [inline third_party/move/move-compiler-v2/src/pipeline/ability_processor.rs:103]: This is the wrong way to fix this. Obviously parameters are not correctly treated by the drop analysis. Instead of adding some new ad-hoc code here which works around all the existing logic, please try to fix the real bug.

Comment (rahxephon89) [reply third_party/move/move-compiler-v2/src/pipeline/ability_processor.rs:103]: Hi @wrwg, I think the issue happens because the drop information for parameters is missing in `CopyDropState` and the error is detected without analyzing the code. Do you think it makes more sense to extend `CopyDropState` with this info and add the check to the Transformer?

**Review by brmataptos — APPROVED** [review]
Please wait for Wolfgang's approval.  I'm less picky about where the checks go.

File: third_party/move/move-compiler-v2/src/pipeline/ability_processor.rs
Comment (rahxephon89) [reply third_party/move/move-compiler-v2/src/pipeline/ability_processor.rs]: I moved the check into the Transformer, which will be called before analyzing the code.

File: third_party/move/move-compiler-v2/src/pipeline/ability_processor.rs
Comment (wrwg) [inline third_party/move/move-compiler-v2/src/pipeline/ability_processor.rs:251]: As I said, we should not reimplement the check and error message a 2nd time. We have this code already, you duplicating a lot of code. The CopyDrop analysis should compute that these drops are needed. It currently does not because livevar info is broken for parameters.

The real bug here is that parameters should be part of the alive.before set _of the first instruction_ of a function. If they aren't used furthermore in the code, they will be automatically dropped by the code here.

So the fix is in livevar not here.

Comment (rahxephon89) [reply third_party/move/move-compiler-v2/src/pipeline/ability_processor.rs:251]: I see your point but if a function parameter is never used in the code, it should not be in the alive.before set of the first instruction because LiveVarInfo.usages can only contain locations corresponding to the instruction, right?

File: third_party/move/move-compiler-v2/src/pipeline/ability_processor.rs
Comment (wrwg) [inline third_party/move/move-compiler-v2/src/pipeline/ability_processor.rs:249]: What is the difference between a function which has some instructions and one which is only one return? The body is not empty in either case.

Comment (rahxephon89) [reply third_party/move/move-compiler-v2/src/pipeline/ability_processor.rs:249]: Sorry, the comment here is confusing. I think the case of only one return should be equivalent with the empty body where drops of parameters needs to be done before the return. If there are multiple instructions, drops of unused parameters can be handled by the existing code.

**Review by wrwg — APPROVED** [review]

**Pull request merged**
